### PR TITLE
Create a generic AVR target: avr-none

### PR DESCRIPTION
**Pull request by Patryk27**
This commit removes the `avr-unknown-gnu-atmega328` target and replaces it with a more generic `avr-none` variant that must be specialized using `-C target-cpu` (e.g. `-C target-cpu=atmega328p`).

Seizing the day, I'm adding myself as the maintainer of this target - I've been already fixing the bugs anyway, might as well make it official 🙂 

Related discussions:
- https://github.com/rust-lang/rust/pull/131171
- https://github.com/rust-lang/compiler-team/issues/800

try-job: x86_64-gnu-debug

### CODE REVIEW

**Comment by Patryk27** [comment]
Ouch, I'm not sure why the test is failing - `rustc --print=all-target-specs-json -Z unstable-options` _does_ return `avr-unknown-unknown`.

**Comment by workingjubilee** [comment]
my first instinct is "bootstrap problem?"

**Comment by jieyouxu** [comment]
I'll take a look

File: src/tools/compiletest/src/header/tests.rs [hunk 1]
```diff
@@ -380,7 +380,7 @@ fn profiler_runtime() {
 #[test]
 fn asm_support() {
     let asms = [
-        ("avr-unknown-gnu-atmega328", false),
+        ("avr-unknown-unknown", false),
```
Comment (jieyouxu) [inline src/tools/compiletest/src/header/tests.rs:383]: The problem is this line, PR CI runs compiletest tool tests at `--stage 0`, which uses a prebuilt beta rustc. This beta rustc does not know about `avr-unknown-unknown` because it doesn't exist to beta rustc, it only knows about `avr-unknown-gnu-atmega328`. You need to guard this with something like

```rs
#[cfg(bootstrap)]
("avr-unknown-gnu-atmega328", false),
#[cfg(not(bootstrap))]
("avr-unknown-unknown", false),
```

Then T-release should delete the `#[cfg(bootstrap)]`'d `avr-unknown-gnu-atmega328` next bump and when a beta rustc knowns about the renamed `avr-unknown-unknown` target delete `#[cfg(not(bootstrap))]` attribute.

**Review by jieyouxu — APPROVED** [review]
The test and bootstrap side of things LGTM modulo the compiletest header test thing. Didn't look at the compiler side.

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+## Requirements
+
+This target is only cross-compiled - x86_64 / aarch64 x Lixux / MacOS hosts are
+confirmed to work, but in principle any machine able to run rustc and avr-gcc
+should be good.
+
+Compiling for this target requires `avr-gcc`, because a couple of intrinsics
+(like 32-bit multiplication) rely on [`libgcc`](https://github.com/gcc-mirror/gcc/blob/3269a722b7a03613e9c4e2862bc5088c4a17cc11/libgcc/config/avr/lib1funcs.S)
+and can't be provided through `compiler-builtins` yet; this is a limitation that
+we hope to lift in the future.
```
Comment (tgross35) [inline src/doc/rustc/src/platform-support/avr-unknown-unknown.md:20]: Mind opening up an issue at https://github.com/rust-lang/compiler-builtins/issues for what exactly these problems are? Might as well make a todo list for the future.

Comment (Patryk27) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md]: Sure, done; I've also linked the issue in this document.

File: src/tools/compiletest/src/header/tests.rs (hunk 1 above)
Comment (Patryk27) [reply src/tools/compiletest/src/header/tests.rs]: Cool, let's try it 🙂

**Comment by tgross35** [comment]
Just checking since I don't see it linked, did this get an MCP? That is usually the approval step for changing targets.

**Comment by Patryk27** [comment]
I haven't created MCP, wasn't sure on the process (especially considering AVR is rather unpopular target).

If you think this needs a MCP, I can create one.

File: compiler/rustc_target/src/spec/tests/tests_impl.rs [hunk 3]
```diff
@@ -39,7 +39,6 @@ impl Target {
             &self.post_link_args,
         ] {
             for (&flavor, flavor_args) in args {
-                assert!(!flavor_args.is_empty());
```
Comment (tgross35) [inline compiler/rustc_target/src/spec/tests/tests_impl.rs:42]: Instead of removing this, could it change to `assert!(!flavor_args.is_empty() || self.arc == "avr");`? I'm not sure exactly why this check is here (looks like @petrochenkov added it) but it seems worth keeping if this target is the only exception.

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+# NixOS (takes a couple of minutes, since Hydra doesn't build it):
+$ nix shell nixpkgs#pkgsCross.avr.buildPackages.gcc11
+```
+
+... setup `.cargo/config` for your project:
+
+```toml
+[build]
+target = "avr-unknown-unknown"
+rustflags = ["-C", "target-cpu=atmega328p"]
```
Comment (tgross35) [inline src/doc/rustc/src/platform-support/avr-unknown-unknown.md:47]: Are there other `target-cpu`s expected to work here?

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+**Tier: 3**
+
+## Target maintainers
+
+-[Patryk Wychowaniec](https://github.com/Patryk27) <pwychowaniec@pm.me>
+
+## Requirements
+
+This target is only cross-compiled - x86_64 / aarch64 x Lixux / MacOS hosts are
```
Comment (tgross35) [inline src/doc/rustc/src/platform-support/avr-unknown-unknown.md:13]: `aarch64 x Lixux` doesn't seem quite right :)

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+Note that since AVRs have rather small amounts of registers, ROM and RAM, it's
+recommended to always use `--release` to avoid running out of space.
+
+## Testing
+
+You can use [`simavr`](https://github.com/buserror/simavr) to emulate the
+resulting firmware on your machine:
+
+```
```
Comment (tgross35) [inline src/doc/rustc/src/platform-support/avr-unknown-unknown.md:70]: Needs ```` ```console ```` for the failing tests

**Comment by tgross35** [comment]
I think an MCP makes sense just to make everyone aware of the change, it's the usual process even for tier 3 targets. I don't see it actually being a problem, this makes sense to me.

Overall I think PR looks pretty good, left a few small comments. Thanks for the thorough build and test instructions, that is awesome to have.

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md (hunk 5 above)
Comment (Patryk27) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md:13]: Hm, why?

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md (hunk 4 above)
Comment (Patryk27) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md:47]: Yes, [quite a bit](https://github.com/llvm/llvm-project/blob/093d4db2f3c874d4683fb01194b00dbb20e5c713/clang/lib/Basic/Targets/AVR.cpp#L32).

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md (hunk 5 above)
Comment (tgross35) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md:13]: Shouldn't it be linux?

The `x` just read kind of weird, maybe reword to something like "x86_64 and aarch64 hosts running MacOS or Linux are confirmed to work..."

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md (hunk 4 above)
Comment (tgross35) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md:47]: Oh, awesome. Could you mention this somewhere in the docs?

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md (hunk 6 above)
Comment (Patryk27) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md]: kk, added

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md (hunk 5 above)
Comment (Patryk27) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md]: Sure, I've adjusted the sentence to make it more readable.

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md (hunk 4 above)
Comment (Patryk27) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md:46]: Sure, added 🙂

File: compiler/rustc_target/src/spec/tests/tests_impl.rs (hunk 3 above)
Comment (Patryk27) [reply compiler/rustc_target/src/spec/tests/tests_impl.rs:42]: kk, changed

**Comment by Patryk27** [comment]
Okie, I'll create an MCP then 🙂

**Comment by Patryk27** [comment]
^ MCP created: https://github.com/rust-lang/compiler-team/issues/800.

**Comment by Patryk27** [comment]
cc @tgross35? 👀

**Comment by madsmtm** [comment]
So, Rust's naming scheme isn't really consistent, but I think the target should be named `avr-unknown-none` (since it has `target_os = "none"`), and then [if you end up wanting variations](https://rust-lang.zulipchat.com/#narrow/channel/233931-t-compiler.2Fmajor-changes/topic/Create.20an.20avr-unknown-unknown.20target.20compiler-team.23800/near/477904333), they should probably be `avr-unknown-none-*`.

There is also precedent for omitting the `unknown` part, examples include `wasm32v1-none`, `msp430-none-elf` and `riscv32imc-esp-espidf`. So that's also an option (calling it `avr-none`, and future variants `avr-none-*`).

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
+# `avr-unknown-unknown`
+
+Series of microcontrollers from Atmel: ATmega8, ATmega328p etc.
+
+**Tier: 3**
+
+## Target maintainers
+
+-[Patryk Wychowaniec](https://github.com/Patryk27) <pwychowaniec@pm.me>
```
Comment (madsmtm) [inline src/doc/rustc/src/platform-support/avr-unknown-unknown.md:9]: Nit:

```suggestion
- [Patryk Wychowaniec](https://github.com/Patryk27) <pwychowaniec@pm.me>
```

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+## Requirements
+
+This target is only cross-compiled; x86\_64 Linux, x86\_64 MacOS and aarch64
+MacOS hosts are confirmed to work, but in principle any machine able to run
```
Comment (madsmtm) [inline src/doc/rustc/src/platform-support/avr-unknown-unknown.md:14]: Nit:

```suggestion
This target is only cross-compiled; x86\_64 Linux, x86\_64 macOS and Aarch64
macOS hosts are confirmed to work, but in principle any machine able to run
```

**Review by madsmtm — APPROVED** [review]

**Comment by tgross35** [comment]
> cc @tgross35? 👀

Somebody needs to second the MCP. The changes make sense to me but I don't know enough about it to make a decision here - I'll poke on Zulip.

**Comment by Flakebi** [comment]
FYI, here is a PR to create an error if no target-cpu is explicitly specified: #135030
The same thing about no default cpu came up in the amdgpu MCP discussion.

**Comment by Patryk27** [comment]
Status: rebasing!

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md (hunk 7 above)
Comment (Patryk27) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md]: kk, done

File: src/doc/rustc/src/platform-support/avr-unknown-unknown.md (hunk 8 above)
Comment (Patryk27) [reply src/doc/rustc/src/platform-support/avr-unknown-unknown.md]: ouch, right! 😄

**Comment by Patryk27** [comment]
Alright, rebased - I've also marked the target as `need_explicit_cpu`, per https://github.com/rust-lang/rust/pull/135030.

File: src/doc/rustc/src/platform-support/avr-none.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+Compiling for this target requires `avr-gcc`, because a couple of intrinsics
+(like 32-bit multiplication) rely on [`libgcc`](https://github.com/gcc-mirror/gcc/blob/3269a722b7a03613e9c4e2862bc5088c4a17cc11/libgcc/config/avr/lib1funcs.S)
+and can't be provided through `compiler-builtins` yet. This is a limitation that
+[we hope to lift in the future](https://github.com/rust-lang/compiler-builtins/issues/711).
+
+## Building the target
+
+Rust comes with AVR support enabled, you don't have to rebuild the compiler.
+
+## Building Rust programs
```
Comment (tgross35) [inline src/doc/rustc/src/platform-support/avr-none.md:26]: Just want to mention that I very much appreciate the detailed instructions for building and emulating, I wish all our targets had this.

File: src/doc/rustc/src/platform-support/avr-none.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+Series of microcontrollers from Atmel: ATmega8, ATmega328p etc.
+
+**Tier: 3**
+
+## Target maintainers
+
+- [Patryk Wychowaniec](https://github.com/Patryk27) <pwychowaniec@pm.me>
+
+## Requirements
```
Comment (tgross35) [inline src/doc/rustc/src/platform-support/avr-none.md:11]: Could you add a note in this section about a target-cpu being required and maybe link to a list? It's in the sample `rustflags` toml config but having it explicit would be nice.

File: tests/run-make/avr-rjmp-offset/rmake.rs [hunk 11]
```diff
@@ -22,7 +22,7 @@ fn main() {
         .input("avr-rjmp-offsets.rs")
         .opt_level("s")
         .panic("abort")
-        .target("avr-unknown-gnu-atmega328")
+        .target("avr-none")
         // normally one links with `avr-gcc`, but this is not available in CI,
         // hence this test diverges from the default behavior to enable linking
         // at all, which is necessary for the test (to resolve the labels). To
```
Comment (tgross35) [inline tests/run-make/avr-rjmp-offset/rmake.rs:28]: How is this working without specifying a target-cpu?

File: src/doc/rustc/src/platform-support.md [hunk 12]
```diff
@@ -300,7 +300,7 @@ target | std | host | notes
 [`armv8r-none-eabihf`](platform-support/armv8r-none-eabihf.md) | * |  | Bare Armv8-R, hardfloat
 [`armv7a-nuttx-eabi`](platform-support/nuttx.md) | ✓ |  | ARMv7-A with NuttX
 [`armv7a-nuttx-eabihf`](platform-support/nuttx.md) | ✓ |  | ARMv7-A with NuttX, hardfloat
-`avr-unknown-gnu-atmega328` | * |  | AVR. Requires `-Z build-std=core`
+`avr-none` | * |  | AVR. Requires `-Z build-std=core`
```
Comment (tgross35) [inline src/doc/rustc/src/platform-support.md:303]: Could you add a note about target-cpu here as well? E.g. https://github.com/rust-lang/rust/blob/56795fb77a69bc61836537175564455067adada6/src/doc/rustc/src/platform-support.md?plain=1#L275

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 13]
```diff
@@ -1986,10 +1986,20 @@ fn add_post_link_objects(
...
+    flavor: LinkerFlavor,
+    codegen_results: &CodegenResults,
+) {
     if let Some(args) = sess.target.pre_link_args.get(&flavor) {
         cmd.verbatim_args(args.iter().map(Deref::deref));
     }
+
+    if sess.target.arch == "avr" {
+        cmd.verbatim_arg(format!("-mmcu={}", codegen_results.crate_info.target_cpu));
+    }
```
Comment (tgross35) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2001]:
```rust
    // avr-none does not have a default ISA, this is specified via target-cpu
```

Or some similar breadcrumb

**Review by tgross35 — APPROVED** [review]
A handful of doc-related nits but overall lgtm. This does need to wait for the MCP to complete though.

File: src/doc/rustc/src/platform-support/avr-none.md [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+This target is only cross-compiled; x86\_64 Linux, x86\_64 macOS and aarch64
```
[the remark this answers is not in the document]
Comment (Patryk27) [reply src/doc/rustc/src/platform-support/avr-none.md]: Okie, applied.

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 13 above)
Comment (Patryk27) [reply compiler/rustc_codegen_ssa/src/back/link.rs]: Sure, added.

File: src/doc/rustc/src/platform-support.md (hunk 12 above)
Comment (Patryk27) [reply src/doc/rustc/src/platform-support.md]: Okie, added.

File: tests/run-make/avr-rjmp-offset/rmake.rs (hunk 11 above)
Comment (Patryk27) [reply tests/run-make/avr-rjmp-offset/rmake.rs:28]: No idea - maybe it's not actually run? (considering the pipelines pass)

Comment (tgross35) [reply tests/run-make/avr-rjmp-offset/rmake.rs:28]: Does it pass locally? If you can rebase, I'll do one of the try jobs that I think has the lld component.

Or maybe try just `panic!()`ing to see if you can force a failure 😆

Comment (tgross35) [reply tests/run-make/avr-rjmp-offset/rmake.rs:28]: Actually yeah, I think the default CI job just doesn't do `needs-llvm-components: avr` since LLVM isn't get built. So if you can rebase without fixing the issue, I'll test a different job so we can verify the failure is what we expect.

**Comment by tgross35** [comment]
I'm inclined to approve this after the [MCP](https://github.com/rust-lang/compiler-team/issues/800) completes (soonish) and we figure out the rmake test. @workingjubilee anything you want to double check here?

File: tests/run-make/avr-rjmp-offset/rmake.rs (hunk 11 above)
Comment (Patryk27) [reply tests/run-make/avr-rjmp-offset/rmake.rs:28]: Alright, I don't think this test makes sense anymore - adding `.arg("-Ctarget-cpu=atmega328p")` causes it to crash with `rust-lld: error: unknown emulation: mcu=atmega328p`, but - at the same time - we can't swap `rust-lld` with the actually expected `avr-gcc` linker, because it's not available on the CI:

https://github.com/rust-lang/rust/blob/8c04e395952022a451138dc4dbead6dd6ae65203/tests/run-make/avr-rjmp-offset/rmake.rs#L26

I'd be up for removing it, unless we can get `avr-gcc` on the CI.

Comment (tgross35) [reply tests/run-make/avr-rjmp-offset/rmake.rs:28]: Ignoring issues with missing symbols, is lld able to link avr binaries correctly if the flags are adjusted? Or was it not working correctly before?

If the former then it would be nicest if this PR could keep it working somehow, so there is less to do once `compiler-builtins` does have the symbols.

Comment (Patryk27) [reply tests/run-make/avr-rjmp-offset/rmake.rs:28]: The issue is that now we're doing `cmd.verbatim_arg(format!("-mmcu={}", codegen_results.crate_info.target_cpu));` within `add_pre_link_args()`, providing `lld` with a flag it doesn't support - and we didn't do that before, because it was a responsibility silently shoved onto `target.json`, which we want to get rid of here, after all.

Considering we have this entire `Linker` abstraction that's available inside `add_pre_link_args()`, I guess we could hack our way around it - e.g. I could add method such as `fn supports_mmcu_flag(&self) -> bool` or whatever, but the only practical use of that would be for this single test (which ain't necessarily a bad thing, ofc.).

Comment (tgross35) [reply tests/run-make/avr-rjmp-offset/rmake.rs:28]: If somebody wanted to use lld in the future that should theoretically be possible right, assuming the symbols came from somewhere? Meaning it's not like lld doesn't support linking for these targets, just not via the same flags as gcc. Or is using lld for avr broken in some way?

I think we should try to not break lld support if at all possible given it should eventually make using the target easier, but I don't know what the best way to get the flags right is. @workingjubilee do you have any ideas?

Comment (Patryk27) [reply tests/run-make/avr-rjmp-offset/rmake.rs:28]: > If somebody wanted to use lld in the future that should theoretically be possible right, assuming the symbols came from somewhere?

Yes, I think so - it is (should be) just the matter of providing appropriate flags.

**Comment by Patryk27** [comment]
Sure! So far I just rebased the code, to keep it fresh, and I'll take a stab at supporting lld in the next days.

**Comment by Patryk27** [comment]
Okie, test fixed!

File: src/tools/run-make-support/src/external_deps/rustc.rs [hunk 15]
```diff
@@ -253,6 +253,13 @@ impl Rustc {
         self
     }
 
+    /// Specify the target CPU.
+    pub fn target_cpu<S: AsRef<str>>(&mut self, target_cpu: S) -> &mut Self {
```
Comment (tgross35) [inline src/tools/run-make-support/src/external_deps/rustc.rs:257]: Cc @jieyouxu for rmake changes (this seems uncontroversial)

Comment (tgross35) [reply src/tools/run-make-support/src/external_deps/rustc.rs:257]: Oh, rustbot did that for me 🙃

**Comment by tgross35** [comment]
Thanks for the updates, the new implementation looks very clean. Added a try job that I believe runs rmake tests, I think this looks good as long as that passes.

@bors try

**Comment by tgross35** [comment]
> (it's cheating a bit, but should be good for the time being)

Do you just mean using `avr` as the target CPU?

**Comment by Patryk27** [comment]
Yeah, that's the cheating part - I couldn't force lld to understand `-mmcu` / `-mcpu` / `--mmcu` / `--mcpu`, so instead I've adjusted the `target-cpu` to the "minimal" variant of just the base instruction set that seems to be lld's default. I'm open to other approaches, just couldn't make it work otherwise myself.

**Comment by tgross35** [comment]
That seems fine to me for the time being. Mind just adding a comment about that in the rmake test?

File: src/tools/run-make-support/src/external_deps/rustc.rs (hunk 15 above)
Comment (jieyouxu) [reply src/tools/run-make-support/src/external_deps/rustc.rs:257]: Yeah that is fine

**Comment by Patryk27** [comment]
Ready!

**Review by tgross35 — APPROVED** [review]

**Comment by tgross35** [comment]
Thanks! Let's get this in.

@bors r+
